src/handlers/tokens/setTokenAuthority.ts

```diff
@@ -0,0 +1,64 @@
+import { TokenClientType } from '@colony/colony-js';
+
+import { ContractEvent } from '~types';
+import {
+  getCachedTokenClient,
+  fetchColoniesByNativeToken,
+  updateSingleColonyNativeTokenStatuses,
+} from '~utils';
+
+export default async (event: ContractEvent): Promise<void> => {
+  const { contractAddress, signature } = event;
+
+  const client = await getCachedTokenClient(contractAddress);
+
+  if (!client) {
+    throw new Error(
+      `Cannot get token client for ${contractAddress} when handling ${signature}`,
+    );
+  }
+
+  const colonies = await fetchColoniesByNativeToken(contractAddress);
+
+  colonies.forEach(async (colony) => {
+    const { id: colonyAddress } = colony;
+    let unlockable = false;
+    let mintable = false;
+
+    // only set the unlock / mint status for colony network tokens
+    if (
+      client.tokenClientType === TokenClientType.Colony ||
+      client.tokenClientType === TokenClientType.ColonyLegacy
+    ) {
+      try {
+        await client.estimateGas.unlock({ from: colonyAddress });
+        unlockable = true;
+      } catch (error) {
+        // silent
+      }
+
+      try {
+        await client.estimateGas['mint(uint256)'](1, { from: colonyAddress });
+        mintable = true;
+      } catch (error) {
+        // silent
+      }
+    } else {
+      // any other type of token
+      // note that we can't be sure if the token even has the mint function,
+      // or if it does, what signature does it have
+      try {
+        // @ts-expect-error
+        await client.estimateGas.mint(1, { from: colonyAddress });
+        mintable = true;
+      } catch (error) {
+        // silent
+      }
+    }
```

Comment:
Will it actually run async if you don't use something like `Promise.all`?

Reply:
You are correct that this might be a problem normally _(even with `Promise.all`)_ as the way `forEach` is designed, it can't handle promises properly _(especially rejections)_

It kinda works for our case, as we don't really care about the result, since each colony is independent and can be updated whenever. 

The real problem here begins if there's ever an error, as `forEach` won't catch it properly and it has the potential of crashing the node process.

As such, I've refactored both this, and the other instance that's using this partern to `for...of`